lib/services/local/events.go

```diff
@@ -352,32 +365,41 @@ func (p *staticTokensParser) parse(event backend.Event) (services.Resource, erro
 	}
 }
 
-func newClusterConfigParser() *clusterConfigParser {
+func newClusterConfigParser(getClusterConfig getClusterConfigFunc) *clusterConfigParser {
+	prefixes := [][]byte{
+		backend.Key(clusterConfigPrefix, generalPrefix),
+		backend.Key(clusterConfigPrefix, networkingPrefix),
+		backend.Key(clusterConfigPrefix, sessionRecordingPrefix),
+	}
 	return &clusterConfigParser{
-		baseParser: baseParser{matchPrefix: backend.Key(clusterConfigPrefix, generalPrefix)},
+		baseParser:       newBaseParser(prefixes...),
+		getClusterConfig: getClusterConfig,
 	}
 }
 
 type clusterConfigParser struct {
 	baseParser
+	getClusterConfig getClusterConfigFunc
 }
 
 func (p *clusterConfigParser) parse(event backend.Event) (services.Resource, error) {
 	switch event.Type {
 	case backend.OpDelete:
+		if !bytes.HasPrefix(event.Item.Key, backend.Key(clusterConfigPrefix, generalPrefix)) {
+			return nil, nil
+		}
 		h, err := resourceHeader(event, services.KindClusterConfig, services.V3, 0)
 		if err != nil {
 			return nil, trace.Wrap(err)
 		}
 		h.SetName(services.MetaNameClusterConfig)
 		return h, nil
 	case backend.OpPut:
-		clusterConfig, err := services.UnmarshalClusterConfig(
-			event.Item.Value,
-			services.WithResourceID(event.Item.ID),
-			services.WithExpires(event.Item.Expires),
-			services.SkipValidation(),
-		)
+		// To ensure backward compatibility, do not use the ClusterConfig
+		// resource passed with the event but perform a separate get from the
+		// backend. The resource fetched in this way is populated with all the
+		// fields expected by legacy event consumers.  DELETE IN 8.0.0
+		clusterConfig, err := p.getClusterConfig()
```

Comment:
this may be breaking the abstraction layers a bit, makes me nervous about cyclical dependencies (in case some code wraps `services.ClusterConfiguration`).
how about extracting `getClusterConfig` into a standalone helper function called from both here and from `GetClusterConfig` to remove a direct dependency between them?

Reply:
I don't quite understand the concern nor the suggested solution.

Reply:
`EventsService` already has direct access to the `backend.Backend`.
theoretically, it shouldn't need an opaque function variable to read the full cluster config.
if there was a `func getClusterConfig(bk backend.Backend) (types.ClusterConfig, error)` in this package, you wouldn't need to pass in a callback.

but this is a temporary kludge, I don't mind keeping it for one release cycle.